src/main/java/com/kovalevskyi/java/deep/model/graph/ConnectedNeuron.java

```diff
@@ -37,32 +39,12 @@
 
     private volatile double forwardInputToActivationFunction;
 
-    public ConnectedNeuron(final Map<Neuron, Double> backwardConnections,
-                           final double bias,
-                           final ActivationFunction activationFunction,
-                           final double learningRate,
-                           final ForkJoinPool forkJoinPool) {
-        this(backwardConnections,
-                bias,
-                activationFunction,
-                learningRate,
-                null,
-                forkJoinPool);
-    }
-
-    public ConnectedNeuron(final Map<Neuron, Double> backwardConnections,
-                           final double bias,
-                           final ActivationFunction activationFunction,
-                           final double learningRate,
-                           final String name,
-                           final ForkJoinPool forkJoinPool) {
-        this.forkJoinPool = forkJoinPool;
-        this.backwardConnections = new ConcurrentHashMap<>(backwardConnections);
-        this.learningRate = learningRate;
-        backwardConnections.keySet().forEach(n -> inputSignals.put(n, Double.NaN));
-        this.bias = bias;
-        this.activationFunction = activationFunction;
-        this.name = name;
+    private ConnectedNeuron(final ConnectedNeuron.Builder builder) {
```

Comment:
Maybe it’s just me, but it feels strange that the constructor is “unpacking” the Builder like this. The `ConnectedNeuron.Builder#build()`method validates its internal state before calling this, yes, but this constructor does no such check.

Reply:
Agree with StrangePan, I'd move validations into constructor

Reply:
Personally, I favor keeping the validation in `Builder.build()` and moving the unpacking there as well, since it reinforces a clear `Builder` -> depends on -> `ConnectedNeuron` relationship. But that’s less important than performing the validation immediatelay before unpacking.

Reply:
Thank you! This is a good observation. I would probably do unpacking in the build method (would prefer not to throw any Exceptions in the constructor).